workspaces/src/types/gas_meter.rs

```diff
@@ -0,0 +1,76 @@
+use std::{
+    ops::AddAssign,
+    sync::{Arc, Mutex},
+};
+
+use crate::Worker;
+
+use super::Gas;
+use crate::result::Result;
+
+/// A hook that is called on every transaction that is sent to the network.
+/// This is useful for debugging purposes, or for tracking the amount of gas
+/// that is being used.
+pub type GasHook = Arc<Mutex<dyn FnMut(Gas) -> Result<()> + Send>>;
+
+/// Allows you to meter the amount of gas consumed by transaction(s).
+/// Note: This only works with parallel transactions that resolve to [`crate::Result::ExecutionFinalResult`]
+/// Example
+/// ```rust, ignore, no_run
+/// let mut worker = workspaces::sandbox().await?;
+/// let meter = GasMeter::now(&mut worker);
+///
+/// let wasm = std::fs::read(STATUS_MSG_WASM_FILEPATH)?;
+/// let contract = worker.dev_deploy(&wasm).await?;
+///
+/// contract
+///    .call("set_status")
+///    .args_json(json!({
+///        "message": "hello_world",
+///    }))
+///    .transact()
+///    .await?;
+///
+/// println!("Total Gas consumed: {}", meter.elapsed()?);
+/// ```
+pub struct GasMeter {
+    gas: Arc<Mutex<Gas>>,
+}
+
+impl GasMeter {
+    /// Create a new gas meter with 0 gas consumed.
+    pub fn now<T: ?Sized>(worker: &mut Worker<T>) -> Self {
+        let mut gas_consumed = Arc::new(Mutex::new(0));
+
+        let meter = Self {
+            gas: Arc::clone(&gas_consumed),
+        };
+
+        worker
+            .on_transact
+            .push(Arc::new(Mutex::new(move |gas: Gas| {
+                // check if reference is unique, expect false
+                match Arc::get_mut(&mut gas_consumed) {
+                    Some(_) => tracing::info!("GasMeter dropped, hook invalid."),
+                    None => gas_consumed.lock()?.add_assign(gas),
+                }
+
+                Ok(())
```

Comment:
The arc reference should not need to be checked like this, so we can just simplify to this:
```suggestion
                gas_consumed.lock()?.add_assign(gas);
                Ok(())
```

Reply:
I wanted to add visibility that the meter is dropped. I made this alternative change:
```rust
 let gas_consumed = Arc::downgrade(&Arc::clone(&meter.gas));
 worker.tx_callbacks.push(Arc::new(move |gas: Gas| {
       // upgrades if meter is still alive, else noop.
      _ = gas_consumed.upgrade().map(|consumed| {
          consumed.lock().expect("meter is valid").add_assign(gas);
      });

      Ok(())
 }));
```